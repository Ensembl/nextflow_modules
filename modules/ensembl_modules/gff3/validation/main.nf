// See the NOTICE file distributed with this work for additional information
// regarding copyright ownership.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

process GFF3_VALIDATION {
    tag "$meta.id"
    label 'process_low'

    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'quay.io/biocontainers/genometools-genometools:1.6.5--py310h3db02ab_0' :
        'biocontainers/genometools-genometools:1.6.5--py310h3db02ab_0' }"    

    input:
        tuple val(meta), path (gene_models, stageAs: "incoming.gff3")

    output:
        tuple val(meta), path("gene_models.gff3"), emit: gene_models
        path "versions.yml", emit: versions

    when:
        task.ext.when == null || task.ext.when

    script:
        out_gff = "gene_models.gff3"
        """
        cp ${gene_models} temp.gff3
        gt gff3 -tidy -sort -retainids -force -o ${out_gff} temp.gff3
        gt gff3validator ${out_gff}
        
        # Get version from gt tools please
        cat <<-END_VERSIONS > versions.yml
        "${task.process}":
            genometools: \$(gt --version | head -1 | sed 's/gt (GenomeTools) //')
        END_VERSIONS
        """

    stub:
        gene_models = "gene_models.gff3"
        """
        echo "No change, create an emply gff" >> ${gene_models}

        # Get version from gt tools please
        cat <<-END_VERSIONS > versions.yml
        "${task.process}":
            genometools: \$(gt --version | head -1 | sed 's/gt (GenomeTools) //')
        END_VERSIONS
        """
}