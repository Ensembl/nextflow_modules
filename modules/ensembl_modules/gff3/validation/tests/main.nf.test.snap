{
    "Gff3 validation": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "gene_models.gff3:md5,30e231754b0254b6b7293ca13e30dc6b"
                    ]
                ],
                "1": [
                    "versions.yml:md5,869e450ac312021d105c5fe1041ee227"
                ],
                "gene_models": [
                    [
                        {
                            "id": "test"
                        },
                        "gene_models.gff3:md5,30e231754b0254b6b7293ca13e30dc6b"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,869e450ac312021d105c5fe1041ee227"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-10-09T14:30:22.143382311"
    }
}